kubeadm: adapt CRI detection and defaults after the dockershim removal

neolit123 opened the pull request:
#### What this PR does / why we need it:

```
kubeadm: do not special case Docker as a container runtime 

crictl already works with the current state of dockershim.
Using the docker CLI is not required and the DockerRuntime
can be removed from kubeadm. This means that crictl
can connect at the dockershim (or cri-dockerd) socket and
be used to list containers, pull images, remove containers, and
all actions that the kubelet can otherwise perform with the socket.

Ensure that crictl is now required for all supported container runtimes
in checks.go. In the help text in waitcontrolplane.go show only
the crictl example.

Remove the check for the docker service from checks.go.
Remove the DockerValidor check from checks.go.
These two checks were special casing Docker as CR and compensating
for the lack of the same checks in dockershim. With the
extraction of dockershim to cri-dockerd, ideally cri-dockerd
should perform the required checks whether it can support
a given Docker config / version running on a host.
```

```
kubeadm: change the default CRI socket to containerd 

Change the default container runtime CRI socket endpoint to the
one of containerd. Previously it was the one for Docker

- Rename constants.DefaultDockerCRISocket to DefaultCRISocket
- Make the constants files include the endpoints for all supported
container runtimes for Unix/Windows.
- Update unit tests related to docker runtime testing.
- In kubelet/flags.go hardcode the legacy docker socket as a check
to allow kubeadm 1.24 to run against kubelet 1.23 if the user
explicitly sets the criSocket field to "npipe:////./pipe/dockershim"
on Windows or "unix:///var/run/dockershim.sock" on Linux.
```

```
kubeadm: update the CRI socket detection logic 

- Throw an error if there is more than one known socket on the host.
- Remove the special handling for docker+containerd.
- Remove the local instances of constants for endpoints for
Windows / Unix and use the defaultKnownCRISockets variable
which is populated from OS specific constants.
- Update error message in detectCRISocketImpl to have more
details.
- Make detectCRISocketImpl accept a list of "known" sockets
- Update unit tests for detectCRISocketImpl and make them
use generic paths such as "unix:///foo/bar.sock".
```

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
first commit is waiting on https://github.com/kubernetes/kubernetes/pull/107314 to merge.
follow up on https://github.com/kubernetes/kubernetes/pull/106973

xref https://github.com/kubernetes/kubeadm/issues/2626

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
ACTION REQUIRED: kubeadm: default the kubeadm configuration to the containerd socket (Unix: unix:///var/run/containerd/containerd.sock, Windows: "npipe:////./pipe/containerd-containerd") instead of the one for Docker. If the "Init|JoinConfiguration.nodeRegistration.criSocket" field is empty during cluster creation and multiple sockets are found on the host always throw an error and ask the user to specify which one to use by setting the value in the field. Make sure you update any kubeadm configuration files on disk, to not include the dockershim socket unless you are still using kubelet version < 1.24 with kubeadm >= 1.24.

Remove the DockerValidor and ServiceCheck for the "docker" service from kubeadm preflight. Docker is no longer special cased during host validation and ideally this task should be done in the now external cri-dockerd project where the importance of the compatibility matters.

Use crictl for all communication with CRI sockets for actions like pulling images and obtaining a list of running containers instead of using the docker CLI in the case of Docker.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

neolit123 committed:
kubeadm: do not special case Docker as a container runtime

crictl already works with the current state of dockershim.
Using the docker CLI is not required and the DockerRuntime
can be removed from kubeadm. This means that crictl
can connect at the dockershim (or cri-dockerd) socket and
be used to list containers, pull images, remove containers, and
all actions that the kubelet can otherwise perform with the socket.

Ensure that crictl is now required for all supported container runtimes
in checks.go. In the help text in waitcontrolplane.go show only
the crictl example.

Remove the check for the docker service from checks.go.
Remove the DockerValidor check from checks.go.
These two checks were special casing Docker as CR and compensating
for the lack of the same checks in dockershim. With the
extraction of dockershim to cri-dockerd, ideally cri-dockerd
should perform the required checks whether it can support
a given Docker config / version running on a host.

neolit123 committed:
kubeadm: change the default CRI socket to containerd

Change the default container runtime CRI socket endpoint to the
one of containerd. Previously it was the one for Docker

- Rename constants.DefaultDockerCRISocket to DefaultCRISocket
- Make the constants files include the endpoints for all supported
container runtimes for Unix/Windows.
- Update unit tests related to docker runtime testing.
- In kubelet/flags.go hardcode the legacy docker socket as a check
to allow kubeadm 1.24 to run against kubelet 1.23 if the user
explicitly sets the criSocket field to "npipe:////./pipe/dockershim"
on Windows or "unix:///var/run/dockershim.sock" on Linux.

neolit123 committed:
kubeadm: update the CRI socket detection logic

- Throw an error if there is more than one known socket on the host.
- Remove the special handling for docker+containerd.
- Remove the local instances of constants for endpoints for
Windows / Unix and use the defaultKnownCRISockets variable
which is populated from OS specific constants.
- Update error message in detectCRISocketImpl to have more
details.
- Make detectCRISocketImpl accept a list of "known" sockets
- Update unit tests for detectCRISocketImpl and make them
use generic paths such as "unix:///foo/bar.sock".

neolit123 commented:
/triage accepted
/priority important-soon
/milestone v1.24

neolit123 commented:
/kind cleanup feature

neolit123 commented:
/cc @SataQiu @pacoxu
PTAL for review

/hold
/approve cancel

SataQiu commented on cmd/kubeadm/app/cmd/config_test.go:
```diff
@@ -366,7 +366,7 @@ func TestImagesPull(t *testing.T) {
 		LookPathFunc: func(cmd string) (string, error) { return "/usr/bin/docker", nil },
```
How about replace `/usr/bin/docker` as `/usr/local/bin/crictl` ?

SataQiu commented on cmd/kubeadm/app/phases/kubelet/flags.go:
```diff
@@ -102,7 +103,13 @@ func buildKubeletArgMapCommon(opts kubeletFlagsOpts) map[string]string {
 	// TODO: https://github.com/kubernetes/kubeadm/issues/2626
 	hasDockershim := opts.kubeletVersion.Major() == 1 && opts.kubeletVersion.Minor() < 24
-	if opts.nodeRegOpts.CRISocket == constants.DefaultDockerCRISocket && hasDockershim {
+	var dockerSocket string
```
Maybe we can just rename `DefaultDockerCRISocket` to `CRISocketDockershim` in `constants_unix.go/constants_windows.go` and mark it as deprecated and only for compatibility?

neolit123 replied:
this commit changes DefaultCRISocket to equal the containerd socket and removes constants_*.go exposure of the dockershim socket with the idea that kubeadm would not know about it after 1.25 and it will be unsupported.

we could keep CRISocketDockershim in constants_*.go for one more release marked as deprecated, but not sure who would use it. for example, i know that minikube forks the kubeadm constants files, but i don't think they use the dockershim socket.

we currently consider the kubeadm files as not importable and we have not support guarantees around publicly exposed constants/functions in kubeadm outside of the API packages.

with that in mind i inlined the socket strings in this function, but if you think it is better to have CRISocketDockershim for one more release i can add it with the same TODO on top of it:
```
// TODO: https://github.com/kubernetes/kubeadm/issues/2626
```

SataQiu replied:
OK, if this is a deliberate attempt to avoid misuse, that is fine. And we can clean this logic in a future release.

afbjorklund replied:
minikube uses `kubeadm` (and `kubectl`) from the kubernetes version to be installed, there is no version skew.

I thought the decision was to **not** have any default, and have CRI "crash" on start the same way that CNI does ?

neolit123 replied:
> I thought the decision was to not have any default, and have CRI "crash" on start the same way that CNI does ?

The default would be visible if the user generates a config file with the 'kubeadm config print*' commands. We call that a static default and it is also used across the codebase for unit tests. It acts like a placeholder / example value.

But on runtime we have the dynamic detection that will try to populate the field (if empty) and error out on multiple or 0 sockets found on the host.

SataQiu commented on cmd/kubeadm/app/phases/kubelet/flags_test.go:
```diff
@@ -142,7 +142,7 @@ func TestBuildKubeletArgMap(t *testing.T) {
 		},
 		{
-			name: "dockershim socket but kubelet version is without built-in dockershim",
+			name: "docker socket but kubelet version is without built-in dockershim",
```
Does the dockershim socket(`/var/run/dockershim.sock`) is same as the docker socket(`/var/run/cri-dockerd.sock`) here?

tao12345666333 replied:
I prefer cri-dockerd

neolit123 replied:
actually there is no need to change the "name" field here. i will revert this.
the tests are going to removed in 1.25.

SataQiu commented on cmd/kubeadm/app/preflight/checks_test.go:
```diff
@@ -920,7 +920,7 @@ func TestImagePullCheck(t *testing.T) {
 		LookPathFunc: func(cmd string) (string, error) { return "/usr/bin/docker", nil },
```
`/usr/bin/docker` -> `/usr/local/bin/crictl`

SataQiu commented on cmd/kubeadm/app/util/runtime/runtime_test.go:
```diff
@@ -252,8 +245,8 @@ func TestPullImage(t *testing.T) {
 		{"invalid: CRI pull error", "unix:///var/run/crio/crio.sock", "image2", true},
-		{"valid: pull image using docker", constants.DefaultDockerCRISocket, "image1", false},
-		{"invalid: docker pull error", constants.DefaultDockerCRISocket, "image2", true},
+		{"valid: pull image using docker", constants.DefaultCRISocket, "image1", false},
+		{"invalid: docker pull error", constants.DefaultCRISocket, "image2", true},
```
Just remove these test cases about docker?

SataQiu commented on cmd/kubeadm/app/util/runtime/runtime_test.go:
```diff
@@ -197,9 +190,9 @@ func TestRemoveContainers(t *testing.T) {
-		{"valid: remove containers using docker", constants.DefaultDockerCRISocket, []string{"k8s_c1", "k8s_c2", "k8s_c3"}, false},
-		{"invalid: docker rm failure", constants.DefaultDockerCRISocket, []string{"k8s_c1", "k8s_c2", "k8s_c3"}, true},
-		{"invalid: docker stop failure", constants.DefaultDockerCRISocket, []string{"k8s_c1", "k8s_c2", "k8s_c3"}, true},
+		{"valid: remove containers using docker", constants.DefaultCRISocket, []string{"k8s_c1", "k8s_c2", "k8s_c3"}, false},
+		{"invalid: docker rm failure", constants.DefaultCRISocket, []string{"k8s_c1", "k8s_c2", "k8s_c3"}, true},
+		{"invalid: docker stop failure", constants.DefaultCRISocket, []string{"k8s_c1", "k8s_c2", "k8s_c3"}, true},
```
Just remove these test cases about docker?

SataQiu commented on cmd/kubeadm/app/util/runtime/runtime_test.go:
```diff
@@ -143,7 +136,7 @@ func TestListKubeContainers(t *testing.T) {
 		{"valid: list containers using CRI socket url", "unix:///var/run/crio/crio.sock", false},
 		{"invalid: list containers using CRI socket url", "unix:///var/run/crio/crio.sock", true},
-		{"valid: list containers using docker", constants.DefaultDockerCRISocket, false},
+		{"valid: list containers using docker", constants.DefaultCRISocket, false},
```
Just remove these test cases about docker?

SataQiu commented:
/retest

SataQiu commented:
A typo in the last commit message?
```
More than one socket now always throws and error and there
is no special handling for docker+containerd.
---
More than one socket now always throws an error and there
is no special handling for docker+containerd.
```

pacoxu commented on cmd/kubeadm/app/cmd/phases/upgrade/node/kubeletconfig.go:
```diff
@@ -102,8 +102,8 @@ func runKubeletConfigPhase() func(c workflow.RunData) error {
 		// https://github.com/kubernetes/kubeadm/issues/2426
-		var nro *kubeadmapi.NodeRegistrationOptions
 		var missingURLScheme bool
+		nro := &kubeadmapi.NodeRegistrationOptions{}
```
This change is merged in #107314.

neolit123 commented:
/cc @fabriziopandini

neolit123 commented:
thanks for the reviews, thus far.
i have addressed the feedback.

during today's kubeadm meeting @fabriziopandini asked whether the change to remove usage of the docker CLI in favor of crictl (1st commit) would work fine.


from my tests this works fine for all that we need in kubeadm.
i have tested it using the following:
- had a very old crictl version (v1.13.0)
- created a 1.23 single node cluster (that still has dockershim in the kubelet)
- called all the `crictl unix:///var/run/dockershim.sock ...` commands to cover the calls that we make here:
https://github.com/kubernetes/kubernetes/blob/936e12c9307d144183f9c3bc21cc39126f1b65fe/cmd/kubeadm/app/util/runtime/runtime.go

from what i saw everything worked, but would appreciate someone double checking it.

SataQiu reviewed (commented):
/lgtm
/retest

pacoxu commented:
For `cri-docker.sock`, should we update the website? 

And there is a discussion about rename `cri-docker.sock` to `cri-dockerd.sock` in https://github.com/Mirantis/cri-dockerd/issues/24.

afbjorklund commented:
> And there is a discussion about rename `cri-docker.sock` to `cri-dockerd.sock` in [Mirantis/cri-dockerd#24](https://github.com/Mirantis/cri-dockerd/issues/24).

The socket path has _already_ been renamed, with the cri-dockerd 0.2.0 release:
https://github.com/Mirantis/cri-dockerd/commit/50c048cb54e52cd9058f044671e309e9fbda82e4

The question is about the systemd units, currently called cri-docker.socket / .service
(But cri-dockerd.socket in the README instructions, which needs to match the files)

neolit123 commented:
updated to change some more defaults to containerd, that i've missed previously (including the API docs):
https://github.com/kubernetes/kubernetes/compare/37aff08ae7dac25af8692756a6ed112939a029e4..f3f1332223667200f85af84cb297c743d72556de

tao12345666333 reviewed (commented):
/lgtm

neolit123 commented:
k/website PR, which should merge after this one:
https://github.com/kubernetes/website/pull/31309

pacoxu reviewed (approved):
ready to merge now.
/unhold

RA489 commented:
/lgtm

neolit123 commented:
/assign @SataQiu
PTAL again for approval.

SataQiu reviewed (approved):
/lgtm
/approve
